File: designer/src/app/core/utils/topological-sort.ts
import { WorkflowNode, WorkflowEdge } from '../models/workflow.types';

interface TopologicalSortResult {
  sorted: WorkflowNode[];
  hasCycle: boolean;
}

/**
 * Performs topological sort on workflow nodes.
 * Returns sorted nodes and whether a cycle was detected.
 */
export function topologicalSort(
  nodes: WorkflowNode[],
  edges: WorkflowEdge[]
): TopologicalSortResult {
  const nodeMap = new Map<string, WorkflowNode>();
  const inDegree = new Map<string, number>();
  const adjacencyList = new Map<string, string[]>();

  // Initialize maps
  nodes.forEach((node) => {
    nodeMap.set(node.id, node);
    inDegree.set(node.id, 0);
    adjacencyList.set(node.id, []);
  });

  // Build adjacency list and calculate in-degrees
  edges.forEach((edge) => {
    const sourceList = adjacencyList.get(edge.source) || [];
    sourceList.push(edge.target);
    adjacencyList.set(edge.source, sourceList);

    const currentInDegree = inDegree.get(edge.target) || 0;
    inDegree.set(edge.target, currentInDegree + 1);
  });

  // Kahn's algorithm for topological sort
  const queue: string[] = [];
  const sorted: WorkflowNode[] = [];

  // Find all nodes with in-degree 0
  inDegree.forEach((degree, nodeId) => {
    if (degree === 0) {
      queue.push(nodeId);
    }
  });

  let processedCount = 0;

  while (queue.length > 0) {
    const nodeId = queue.shift()!;
    const node = nodeMap.get(nodeId);
    if (node) {
      sorted.push(node);
      processedCount++;
    }

    const neighbors = adjacencyList.get(nodeId) || [];
    neighbors.forEach((neighborId) => {
      const currentInDegree = inDegree.get(neighborId) || 0;
      inDegree.set(neighborId, currentInDegree - 1);

      if (inDegree.get(neighborId) === 0) {
        queue.push(neighborId);
      }
    });
  }

  // If processed count is less than total nodes, there's a cycle
  const hasCycle = processedCount < nodes.length;

  return { sorted, hasCycle };
}

File: api/src/main/java/com/workflow/api/util/TopologicalSort.java
package com.workflow.api.util;

import com.workflow.api.dto.WorkflowEdge;
import com.workflow.api.dto.WorkflowNode;
import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.*;

public class TopologicalSort {
    
    @Data
    @AllArgsConstructor
    public static class Result {
        private List<WorkflowNode> sorted;
        private boolean hasCycle;
    }

    public static Result sort(List<WorkflowNode> nodes, List<WorkflowEdge> edges) {
        Map<String, WorkflowNode> nodeMap = new HashMap<>();
        Map<String, Integer> inDegree = new HashMap<>();
        Map<String, List<String>> adjacencyList = new HashMap<>();

        // Initialize maps
        for (WorkflowNode node : nodes) {
            nodeMap.put(node.getId(), node);
            inDegree.put(node.getId(), 0);
            adjacencyList.put(node.getId(), new ArrayList<>());
        }

        // Build adjacency list and calculate in-degrees
        for (WorkflowEdge edge : edges) {
            List<String> sourceList = adjacencyList.get(edge.getSource());
            sourceList.add(edge.getTarget());
            adjacencyList.put(edge.getSource(), sourceList);

            inDegree.put(edge.getTarget(), inDegree.get(edge.getTarget()) + 1);
        }

        // Kahn's algorithm for topological sort
        Queue<String> queue = new LinkedList<>();
        List<WorkflowNode> sorted = new ArrayList<>();

        // Find all nodes with in-degree 0
        for (Map.Entry<String, Integer> entry : inDegree.entrySet()) {
            if (entry.getValue() == 0) {
                queue.offer(entry.getKey());
            }
        }

        int processedCount = 0;

        while (!queue.isEmpty()) {
            String nodeId = queue.poll();
            WorkflowNode node = nodeMap.get(nodeId);
            if (node != null) {
                sorted.add(node);
                processedCount++;
            }

            List<String> neighbors = adjacencyList.get(nodeId);
            if (neighbors != null) {
                for (String neighborId : neighbors) {
                    int currentInDegree = inDegree.get(neighborId);
                    inDegree.put(neighborId, currentInDegree - 1);

                    if (inDegree.get(neighborId) == 0) {
                        queue.offer(neighborId);
                    }
                }
            }
        }

        // If processed count is less than total nodes, there's a cycle
        boolean hasCycle = processedCount < nodes.size();

        return new Result(sorted, hasCycle);
    }
}
